services/worker-manager/src/data.js

```diff
@@ -46,41 +55,171 @@ WorkerType.prototype.serializable = function() {
+
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+  ${description}
+
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+  ${extraInfo}
+      `.trim(),
     });
-    wt.errors = wt.errors.slice(0, 100);
+  }
+  await WorkerTypeError.create({
+    workerType: this.name,
+    errorId: slugid.v4(),
+    reported: new Date(),
+    kind,
+    title,
+    description,
+    extra,
   });
-  await notify.email({
-    address: owner,
-    subject: `Taskcluster Worker Manager Error: ${title}`,
-    content: `
-Worker Manager has encountered an error while trying to provision the workertype ${this.name}:
-
-\`\`\`
-${description}
-\`\`\`
-
-It includes the extra information:
-
-\`\`\`json
-${yaml.safeDump(extra)}
-\`\`\`
-    `.trim(),
+};
+
+const WorkerTypeError = Entity.configure({
+  version: 1,
+  partitionKey: Entity.keys.StringKey('workerType'),
+  rowKey: Entity.keys.StringKey('errorId'),
+  properties: {
+    // The workertype this maps to.
+    workerType: Entity.types.String,
+
+    // An arbitrary id for this error
+    errorId: Entity.types.SlugId,
+
+    // The datetime this error occured
+    reported: Entity.types.Date,
+
+    // The sort of error this is. Can be used by UIs to differentiate
+    kind: Entity.types.String,
+
+    // A human readable name for this error
+    title: Entity.types.String,
+
+    // A human readable description of this error and what can be done to fix it
+    description: Entity.types.String,
+
+    // Anything else that a reporter may want to add in a structured way
+    extra: Entity.types.JSON,
+  },
+});
+
+WorkerTypeError.prototype.serializable = function() {
+  return {
+    workerType: this.workerType,
+    errorId: this.errorId,
+    reported: this.reported.toJSON(),
+    kind: this.kind,
+    title: this.title,
+    description: this.description,
+    extra: this.extra,
+  };
+};
+
+WorkerTypeError.expire = async (threshold) => {
+  await this.scan({
+    reported: Entity.op.lessThan(threshold),
+  }, {
+    limit: 500,
+    handler: async item => {
+      await item.remove();
+    },
+  });
+};
+
+const Worker = Entity.configure({
+  version: 1,
+  partitionKey: Entity.keys.StringKey('workerType'),
+  rowKey: Entity.keys.StringKey('workerId'),
+  properties: {
+    // The workertype this maps to.
+    workerType: Entity.types.String,
+
+    // The id of this worker
+    workerId: Entity.types.String,
+
+    // The provider responsible for this worker
+    provider: Entity.types.String,
+
+    // The time that this worker was created
+    created: Entity.types.Date,
+
+    // The time that this worker is no longer needed and
+    // should be deleted
+    expires: Entity.types.Date,
+
+    // A string specifying the state this worker is in
+    // so far as worker-manager knows. This can be any
+    // of the fields defined in the enum below.
+    state: Entity.types.String,
+
+    // Anything a provider may want to remember about this worker
+    providerData: Entity.types.JSON,
```

Comment:
will this include some info on the resources (like instance IDs and whatnot?)

Reply:
Not instanceIds currently. It probably makes sense to store that there rather than reconstructing it from the workerId every time by appending or removing `gcp-`.

Reply:
We should probably ensure this data structure has a finite maximum size -- that is, not arrays of arbitrary length.  Otherwise we'll eventually hit the size limit for a single Azure entity, leading to great sadness.